File: Week_04/Seminar/Solutions/OddEvenLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(!head || !head->next) {
            return head;
        }
        
        ListNode* iter = head;
        ListNode* odd = head;
        ListNode* even = nullptr;
        ListNode* evenHead = nullptr;
        
        size_t index = 1;
        while(iter) {
            auto current = iter;
            iter = iter->next;
            if(index & 1) {
                odd->next = current;
                odd = odd->next;
            }
            else {
                if(!evenHead) {
                    even = current;
                    evenHead = current;
                }
                else {
                    even->next = current;
                    even = even->next;
                }
            }
            index++;
        }
        odd->next = evenHead;
        even->next = nullptr;
        return head;
    }
};

File: Week_05/Practicum/PythonAdditional/Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        current = head
        first_even = head.next

        last_odd = None

        i = 1
        while current.next:
            next_node = current.next

            current.next = current.next.next

            if current.next is None and i % 2 == 1:
                last_odd = current

            current = next_node
            i += 1

        last_odd = current if not last_odd else last_odd

        last_odd.next = first_even

        return head